File: Medium/1358. Number of Substrings Containing All Three Characters.py
from collections import defaultdict


class Greedy:
    def numberOfSubstrings(self, s: str) -> int:
        pos = [-1, -1, -1]
        res = 0
        for i, c in enumerate(s):
            pos[ord(c) - ord('a')] = i
            res += min(pos) + 1
        return res

class HashTable:
    def numberOfSubstrings(self, s: str) -> int:
        cnt = defaultdict(int)
        l, n, res = 0, len(s), 0
        for r in range(n):
            cnt[s[r]] += 1
            while cnt['a'] and cnt['b'] and cnt['c']:
                res += n - r
                cnt[s[l]] -= 1
                l += 1
        return res

File: Medium/1358. Number of Substrings Containing All Three Characters.java
class Solution {
    public int numberOfSubstrings(String s) {
        int a = -1, b = -1, c = -1;
        int res = 0;
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (ch == 'a') {
                a = i;
            } else if (ch == 'b') {
                b = i;
            } else {
                c = i;
            }
            res += Math.min(a, Math.min(b, c)) + 1;
        }
        return res;
    }
}
